### ISSUE TITLE
'No route to host?' with active Internet connection

### ISSUE THREAD

**Issue by Filazapovich** [defold/editor2-issues#2746]
Part of people (looks like ~10%) whose download our game is affected by strange error with active and stable Internet connection:
```
ERROR:SCRIPT: Unable to create HTTP connection to 'https://srv.gizmo.dev1.fa-dev.gke.mgsn.it/auth/Register'. No route to host?
```

Native **gethostbyname** works fine and resolves host IP. Logcat from device: 
```
D/libc-netbsd(17157): [getaddrinfo]: mtk hostname=srv.gizmo.dev1.fa-dev.gke.mgsn.it; servname=(null); cache_mode=(null), netid=0; mark=0
D/libc-netbsd(17157): getaddrinfo( app_uid:10242
D/libc-netbsd(17157): getaddrinfo() uid prop:
D/libc-netbsd(17157): getaddrinfo() getuid():10242
D/libc-netbsd(17157): [getaddrinfo]: mtk ai_addrlen=0; ai_canonname=(null); ai_flags=0; ai_family=0
D/libc-netbsd(  356): [getaddrinfo]: mtk hostname=srv.gizmo.dev1.fa-dev.gke.mgsn.it; servname=(null); cache_mode=local, netid=122; mark=917626
D/libc-netbsd(  356): getaddrinfo( app_uid:0
D/libc-netbsd(  356): [getaddrinfo]: mtk ai_addrlen=0; ai_canonname=(null); ai_flags=0; ai_family=0
D/libc-netbsd(  356): res_queryN name = srv.gizmo.dev1.fa-dev.gke.mgsn.it, class = 1, type = 28
D/libc-netbsd(  356): res_queryN name = srv.gizmo.dev1.fa-dev.gke.mgsn.it, class = 1, type = 1
D/libc-netbsd(  356): res_queryN name = srv.gizmo.dev1.fa-dev.gke.mgsn.it succeed
D/libc-netbsd(17157): getaddrinfo: srv.gizmo.dev1.fa-dev.gke.mgsn.it get result from proxy >>
```

Thinks about than:
- That is not problem with Internet because browser and other applications works fine.
- `http.request` uses c-ares lib to resolve host IP. Looks like case with that lib? https://github.com/nodejs/node/issues/17850
- We can partially reproduce this behavior by https://play.google.com/store/apps/details?id=jp.co.taosoftware.android.packetcapture. But some others application doesn't work properly with this VPN.

Our QA have 2 devices with that error: Sony Xperia M5 5603 with Android 5.1 and Lenovo Vibe K5 Plus A6020a46 with Android 5.1.1. That is not case with Android 5.1.

Reproducible with defold 161, 159 and earlier.

<hr/>

<table>
<tr><td>Build time</td><td>2019-09-02T09:40:18.383211</td></tr>
<tr><td>Defold channel</td><td>stable</td></tr>
<tr><td>Defold editor sha</td><td>45635ad26f85009c52905724e242cc92dd252146</td></tr>
<tr><td>Defold engine sha</td><td>45635ad26f85009c52905724e242cc92dd252146</td></tr>
<tr><td>Defold version</td><td>1.2.161</td></tr>
<tr><td>GPU</td><td>Intel(R) HD Graphics 630</td></tr>
<tr><td>GPU Driver</td><td>4.5.0 - Build 23.20.16.4973</td></tr>
<tr><td>Java version</td><td>11.0.1+13</td></tr>
<tr><td>OS arch</td><td>amd64</td></tr>
<tr><td>OS name</td><td>Windows 10</td></tr>
<tr><td>OS version</td><td>10.0</td></tr>
</table>

**Comment by britzl** [comment]
Thank you for reporting this. We haven't heard of this being an issue for anyone else but never the less we are currently investigating it.

**Comment by jhonnyking** [comment]
"Lenovo Vibe K5 Plus A6020a46 with Android 5.1.1. That is not case with Android 5.1."
Does this mean that you can't reproduce it on Android 5.1 with that device?
Does the issue only happen on Android, or do you have similar issues on iOS?

I took a look into the link you provided, and i I understand correctly the addition that was made was that you can provide c_ares with a list of search locations that it should look through for the resolv.conf file. This has since been adapted into c_ares, but it seems strange that the default implementation cannot find the resolve file. Could you look into the file system on those devices and see if the /etc/resolv.conf file exists? and possibly if you have read access to it and what its content is.

**Comment by Filazapovich** [comment]
> "Lenovo Vibe K5 Plus A6020a46 with Android 5.1.1. That is not case with Android 5.1."
> Does this mean that you can't reproduce it on Android 5.1 with that device?
> Does the issue only happen on Android, or do you have similar issues on iOS?

This case is 100% reproducible on Sony Xperia M5 5603 and on Lenovo Vibe K5 Plus A6020a46 from our QA device bank. Both with Android 5.1 but on other devices with Android 5.1 this error is not reproducible. I have no information about similar issue on iOS.

**Comment by jhonnyking** [comment]
I was able to run into a few issues with tpacketcapture app. It does fail on android 8+ devices 100% of the time, but only occasionally on the 5.1 devices:

* When it fails for the 5.1 devices, we get ssl handshake failure after 300-1500 sequential requests. If I switch the request to regular http instead, it doesn't seem to fail at all. Yesterday I could do 60k+ sequential requests without any of them failing.
* For the 8+ devices, we don't get any DNS servers at all when running in capture mode, so it's not surprising that it fails. In this case we discussed adding a default DNS server such as 8.8.8.8 and see if it helps.

So for now, until we can patch the DNS case, could you switch your request from https to http and see if it works better? Obviously not ideal, but consider it a workaround for now. I'll see if I can send you a WIP engine that you guys can try out this week.

**Comment by Filazapovich** [comment]
We made fallback to IP as temporary solution.

**Comment by Filazapovich** [comment]
I think the better solution of no DNS servers problem is property in game.project [network] section with comma separated list of DNS servers used as default servers if no DNS servers found on system. And default value for this property should be 8.8.8.8. This solution make no problems with China distribution.

**Comment by jhonnyking** [comment]
I have a PR that sets DNS servers to 8.8.8.8 and 8.8.4.4 (and similar IPV6 variants) if ares didn't find anything from the device configuration, so if you want I could send you that version so you can test it and see if it resolves your issue. I have tested it on a device with android 8 and the tpacketcapture app running and there are no more DNS issues.

**Comment by Filazapovich** [comment]
Ok. I'm ready to test.

**Comment by jhonnyking** [comment]
osx: https://d.defold.com/archive/f9ab130e3b955b5606202b6b883c5e34a32ef560/stable/editor2/Defold-x86_64-darwin.dmg
win32: https://d.defold.com/archive/f9ab130e3b955b5606202b6b883c5e34a32ef560/stable/editor2/Defold-x86_64-win32.zip
linux: https://d.defold.com/archive/f9ab130e3b955b5606202b6b883c5e34a32ef560/stable/editor2/Defold-x86_64-linux.zip

**Comment by Filazapovich** [comment]
Test failed. I give more info in Slack

**Comment by britzl** [comment]
Will investigate.

**Comment by britzl** [comment]
Surprisingly the https request to graph.facebook.com/me fails with a socket error. The -78 is MBEDTLS_ERR_NET_SEND_FAILED. Doing the same with curl gives a response:

    {"error":{"message":"An active access token must be used to query information about the current user.","type":"OAuthException","code":2500,"fbtrace_id":"Ai5KpDDPg2Pdeu-APCy2kPr"}}

We must be missing something in our config or in the way we set up timeouts

**Comment by britzl** [comment]
Correction and some additional observations:

The socket error doesn't show up 100% of the time (more like 50%).

The test provided in the example makes both GET and POST requests at the same time. If I only do GET requests the problem never happens.

If I do only a POST request the problem happens.

**Comment by britzl** [comment]
Isolating the problem further:

	http.request("https://graph.facebook.com/me", "POST", function(self, id, response)
		print("POST : " .. response.status)
	end, nil, "")

Sometime results in:

    WARNING:DLIB: Unhandled ssl status code: -78 (-FFFFFFB2)

Aaaand it never happens if the request actually contains POST data:

	http.request("https://graph.facebook.com/me", "POST", function(self, id, response)
		print("POST : " .. response.status)
	end, nil, "foo=bar")

This makes sense from one point of view, but I guess a POST request should contain data.

**Comment by britzl** [comment]
We will include a fix for 0 content length POSTs on SSL connections in 1.2.165.

**Comment by britzl** [comment]
@Filazapovich The fix was included in 1.2.164. Can you please confirm that you no longer have any problems with HTTP(S) requests?

**Comment by Filazapovich** [comment]
We will update Defold to 1.2.164+ only at January.
